### PR TITLE
SUT response + annotation pipeline

### PR DESCRIPTION
**Pull request by bkorycki**
Merges `run-prompts` and `run-annotators` commands into `run-csv-items`.

- `run-csv-items -s sut_1 -s sut_2 prompts.csv`: Get the suts' responses to the prompts the in the input file. 
    - Input CSV headers: UID, Text
    - Output (CSV) headers: UID, Text, sut_1, sut_2
      - SUT columns contain responses
- `run-csv-items -s sut_1 -s sut_2 -a annotator_1 -a annotator_2 prompts.csv` Get the suts' responses to the prompts and then annotate the prompt + responses.
    - Input CSV headers: UID, Text
    - Output: Same as the output below.
- `run-csv-items -a annotator_1 -a annotator_2 prompt_responses.csv` Get annotations for prompt + responses in the input file.
    - Input CSV headers: UID, Prompt, SUT, Response
    - Output: JSON lines file. Each json object has the form:
        ```
        {
          "UID": ..., "Prompt": ..., "SUT": ..., "Response": ..., 
          "Annotations": {"annotator_1": ..., "annotator_2": ...}
        {
       ```

### CODE REVIEW

File: modelgauge/main.py [hunk 1]
```diff
@@ -280,123 +287,82 @@ def run_test(
 )
 @click.argument(
     "filename",
-    type=click.Path(exists=True),
+    type=click.Path(exists=True, path_type=pathlib.Path),
 )
-def run_prompts(sut_uids, workers, cache_dir: pathlib.Path, debug, filename):
-    """Take a CSV of prompts and run them through SUTs. The CSV file must have UID and Text columns, and may have others."""
-
+def run_prompts(sut_uids, annotator_uids, workers, sut_cache_dir, debug, filename):
```
Comment (wpietri) [inline modelgauge/main.py:292]: This function is long and hard to follow. Could it be broken up some, either by extracting functions or by creating some sort of state object or runner object?

File: modelgauge/prompt_pipeline.py [hunk 2]
```diff
@@ -1,14 +1,25 @@
...
 
 from modelgauge.pipeline import Source, Pipe, Sink, CachingPipe
 from modelgauge.prompt import TextPrompt
 from modelgauge.single_turn_prompt_response import PromptWithContext
-from modelgauge.sut import PromptResponseSUT, SUT
+from modelgauge.sut import PromptResponseSUT, SUT, SUTCompletion
+
+
+@dataclass
+class SutInteraction:
```
Comment (wpietri) [inline modelgauge/prompt_pipeline.py:16]: Nice. Typing helps clarify things, and I like the name especially.

File: tests/test_annotation_pipeline.py [hunk 3]
```diff
@@ -182,3 +197,47 @@ def track_progress(data):
 
     assert progress_items[0]["completed"] == 0
     assert progress_items[-1]["completed"] == 4
+
+
+@pytest.mark.parametrize(
+    "sut_worker_count,annotator_worker_count",
+    [(1, 1), (2, 2), (8, 8), (1, 5), (5, 1), (3, 9), (9, 3)],
```
Comment (wpietri) [inline tests/test_annotation_pipeline.py:204]: I like this too.

File: tests/test_cli.py [hunk 4]
```diff
@@ -92,7 +96,7 @@ def test_run_prompts_normal(tmp_path):
     assert result.exit_code == 0
 
     out_path = re.findall(r"\S+\.csv", result.stdout)[0]
-    with open(in_path.parent / out_path, "r") as f:
+    with open(tmp_path / out_path, "r") as f:
```
Comment (wpietri) [inline tests/test_cli.py:99]: Ah, nice cleanup. Thanks.

**Review by wpietri — CHANGES_REQUESTED** [review]
Looks solid except for the CLI method being quite large and hard to follow.

File: modelgauge/main.py (hunk 1 above)
Comment (bkorycki) [reply modelgauge/main.py]: That's fair, it was pretty long. I just updated it. Does it look better now?

File: modelgauge/prompt_pipeline.py (hunk 2 above)
Comment (bkorycki) [reply modelgauge/prompt_pipeline.py:16]: Yay!

File: modelgauge/main.py (hunk 1 above)
Comment (wpietri) [reply modelgauge/main.py]: I think this is getting closer, but I think it's still confusing. Some of what I think of as the CLI work, like validating arguments, printing user messages, and progress bars, is now off in another file. The functional decomposition with descriptive names does help convey more of the intent. But the functions have a lot of arguments, many in common. And they have a fair bit of branching, including spots with similar code, like in the `if run_annotators` block.

Instead of a functional breakdown, maybe a PipelineRunner object would help? Happy to sketch something if you'd like.

**Review by wpietri — CHANGES_REQUESTED** [review]
I guess for GH ritual reasons I should say: good progress, needs to go further, please see the long comment that I perhaps should have put here.

File: modelgauge/main.py (hunk 1 above)
Comment (bkorycki) [reply modelgauge/main.py]: Do you imagine `PipelineRunner` to handle all 3 types of pipelines?

Comment (wpietri) [reply modelgauge/main.py]: I'm thinking so. When you have a bunch of things that keep appearing in function signatures together, that's a sign that there might be an object waiting to be named, where those are the fields of the object. When there is code that goes with them, that could end up as methods on the object. Seems like there's some common stuff to all three pipelines!

Comment (bkorycki) [reply modelgauge/main.py]: Okay, I put the logic into a runner object. What do you think?

**Review by wpietri — APPROVED** [review]
Great work! This seems much improved to me. I can now easily tell which bits go with which case, and I like the separation of concerns between the runner object and the CLI itself. How does it feel to you?
